config/loader_test.go

```diff
@@ -0,0 +1,260 @@
+	assert.Equal(t, "Intergalactic Delivery Boy", fry.Occupation)
+	assert.Equal(t, 1025, fry.Age)
+}
+
+func Test_Loader_load_from_viper_nil(t *testing.T) {
+	var person person
+
+	loader := NewLoader(&person)
+
+	err := loader.LoadFromViper(nil)
+	assert.ErrorContains(t, err, "nil configuration")
+}
+
+func Test_Loader_extra_value(t *testing.T) {
+	input := map[string]interface{}{
+		"name":       "Turanga Leela",
+		"occupation": "captain",
+		"age":        24,
+		"numEyes":    1,
+	}
+
+	res := person{}
+	loader := NewLoader(&res)
+	require.NotNil(t, loader)
+
+	err := loader.LoadFromMap(input)
+	assert.ErrorContains(t, err, "unexpected directives: numEyes")
+
+	neloader := NewNonExclusiveLoader(&res)
+	err = neloader.LoadFromMap(input)
+	assert.NoError(t, err)
+}
+
+func Test_Loader_missing_value(t *testing.T) {
+	input := map[string]interface{}{
+		"name": "Turanga Leela",
+		"age":  24,
+	}
+
+	res := person{}
+	loader := NewLoader(&res)
+	require.NotNil(t, loader)
+
+	err := loader.LoadFromMap(input)
+	assert.ErrorContains(t, err, "directives not found: Occupation")
+}
+
+func Test_Loader_invalid_value(t *testing.T) {
+	input := map[string]interface{}{
+		"name":       "Philip J. Fry",
+		"occupation": "own grandfather",
+		"age":        -28,
+	}
+
+	res := person{}
+	loader := NewLoader(&res)
+	require.NotNil(t, loader)
+
+	err := loader.LoadFromMap(input)
+	assert.ErrorContains(t, err, "age cannot be negative")
+}
+
+func Test_Loader_zero_values(t *testing.T) {
+	res := person{
+		Name:       "Philip J. Fry",
+		Occupation: "age-reducing tar victim",
+		Age:        0,
+	}
+	loader := NewLoader(&res)
+	require.NotNil(t, loader)
+
+	// Default zero value is semantically identical to an unset value.
+	err := loader.LoadFromMap(map[string]interface{}{})
+	assert.ErrorContains(t, err, "directives not found: Age")
+
+	// Zero values in the input are ok, however.
+	err = loader.LoadFromMap(map[string]interface{}{"age": 0})
+	assert.NoError(t, err)
+
+	// Setting `config:"zerodefault"` tag on a field propagates zero-value defaults.
+	ageless := struct {
+		Name       string
+		Occupation string `config:"zerodefault"`
+		Age        int    `config:"zerodefault"`
+	}{Name: "Philip J. Fry", Occupation: "", Age: 0}
+
+	loader = NewLoader(&ageless)
+	require.NotNil(t, loader)
+	err = loader.LoadFromMap(map[string]interface{}{})
+	assert.NoError(t, err)
+}
+
+func Test_Loader_renamed_field(t *testing.T) {
+	leela := struct {
+		Name         string
+		Species      string
+		NumberOfEyes int `mapstructure:"number-of-eyes"`
+	}{}
+
+	loader := NewLoader(&leela)
+	require.NotNil(t, loader)
+
+	input := map[string]interface{}{
+		"name":           "Turanga Leela",
+		"species":        "cyclops",
+		"number-of-eyes": 1,
+	}
+
+	err := loader.LoadFromMap(input)
+	require.NoError(t, err)
+	assert.Equal(t, 1, leela.NumberOfEyes)
+}
+
+func Test_Loader_valid_tags(t *testing.T) {
+
+	cfg := struct {
+		ServerAddress string `valid:"dialstring" mapstructure:"addr"`
```

Comment:
the potential use of off-the-shelf govalidator thingies is super cool, and I guess worth documenting in the README.